src/Service.js

```diff
@@ -376,26 +376,33 @@ Service_.prototype.handleCallback = function(callbackRequest) {
  *     otherwise.
  */
 Service_.prototype.hasAccess = function() {
-  return this.lockable_(function() {
-    var token = this.getToken();
-    if (!token || this.isExpired_(token)) {
-      try {
-        if (token && this.canRefresh_(token)) {
-          this.refresh();
-        } else if (this.privateKey_) {
-          this.exchangeJwt_();
-        } else if (this.grantType_) {
-          this.exchangeGrant_();
-        } else {
+  var token = this.getToken();
+  if (!token || this.isExpired_(token)) {
```

Comment:
An improvement to avoid locks when we don't have a token and can't get a new one:

```
var token = this.getToken();
var canGetToken = false;
if ( token && this.isExpired_(token)) {
// I can get a token if I have a refresh_token
canGetToken = token.refresh_token;
} else if ( token ) {
// I have a unexpired token
return true;
} else {
// I don't have a token but maybe I can get one if...
canGetToken = this.privateKey_ || this.grantType_;
}
if ( ! canGetToken  ) {
return false;
}
// I may refresh the token or get one
return this.lockable_(function() {...}),
```

Thoughts?

Reply:
(I've formatted the code above. I'll let Eric comment)

Reply:
Good idea. I've used a slightly different logic, but the same idea of skipping the locking if there is no chance of getting a token.